### PR TITLE
fix(Scripts/AllCommandScript) CanExecuteCommand renamed and moved to be executed for all commands.

### PR DESCRIPTION
**Pull request by AnchyDev**
## Changes Proposed:

This PR proposes changes to:
-  [ ] Core (units, players, creatures, game systems).
-  [x] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

When using the AllCommands::CanExecuteCommand hook, only failed commands are captured.
This is because the call is inside of the failure routine, so I moved it.

## Issues Addressed:

None

## Tests Performed:

This PR has been:
- [x] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

1. Set a breakpoint on line 328 of ChatCommand.cpp
2. Observe it is hit for all commands, failed or not.
3. Yay?

Alternatively you can download [this](https://github.com/AnchyDev/ButWhy) module and observe that only failed commands are blocked.

## Known Issues and TODO List:

I don't see any issue with capturing all commands.

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Chat/ChatCommands/ChatCommand.cpp [hunk 1]
```diff
@@ -322,8 +322,14 @@ namespace Acore::Impl::ChatCommands
         oldTail = newTail;
     }
 
+    /* if we matched a command at some point, invoke it */
     if (cmd)
-    { /* if we matched a command at some point, invoke it */
+    {
+        if (!sScriptMgr->CanExecuteCommand(handler, cmdStr))
+        {
```
Comment (sudlud) [inline src/server/game/Chat/ChatCommands/ChatCommand.cpp:329]: You can remove the braces

Comment (AnchyDev) [reply src/server/game/Chat/ChatCommands/ChatCommand.cpp:329]: > You can remove the braces

Alright, it looks like the AzerothCore code style changed lol. I'll update it

File: src/server/game/Chat/ChatCommands/ChatCommand.cpp [hunk 2]
```diff
@@ -345,9 +344,7 @@ namespace Acore::Impl::ChatCommands
     }
 
     if (!sScriptMgr->CanExecuteCommand(handler, cmdStr))
```
Comment (sudlud) [inline src/server/game/Chat/ChatCommands/ChatCommand.cpp:346]: Why are we having this twice here? 👀

**Review by sudlud — COMMENTED** [review]
Why calling the hook twice

File: src/server/game/Chat/ChatCommands/ChatCommand.cpp (hunk 2 above)
Comment (AnchyDev) [reply src/server/game/Chat/ChatCommands/ChatCommand.cpp:346]: I don't know, it was originally there so I just left it. It can probably be removed.

**Comment by AnchyDev** [comment]
> Why calling the hook twice

Monkeys

**Comment by sudlud** [comment]
looking again at this, it feels like this hook was specifically placed to be called only on unknown / failed command executions.
looks like it was introduced with #9281 and replacing `sEluna->OnCommand(...)`, so there is at least a naming difference which implies a different use case

**Review by sudlud — APPROVED** [review]

**Pull request merged**